src/hotspot/share/opto/divnode.cpp

```diff
@@ -504,6 +504,98 @@ Node* unsigned_div_ideal(PhaseGVN* phase, bool can_reshape, Node* div) {
+  int widen = MAX2(i1->_widen, i2->_widen);
+
+  // Case A: divisor range spans zero (i2->_lo < 0 < i2->_hi)
+  // We split into two subproblems to avoid division by 0:
+  //   - negative part: [i2->_lo, −1]
+  //   - positive part: [1, i2->_hi]
+  // Then we union the results by taking the min of all lower‐bounds and
+  // the max of all upper‐bounds from the two halves.
+  if (i2->_lo < 0 && i2->_hi > 0) {
+    // Handle negative part of the divisor range
+    const IntegerType* neg_part = compute_signed_div_type(i1, IntegerType::make(i2->_lo, -1, widen));
+    // Handle positive part of the divisor range
+    const IntegerType* pos_part = compute_signed_div_type(i1, IntegerType::make(1, i2->_hi, widen));
+    // Merge results
+    NativeType new_lo = MIN2(neg_part->_lo, pos_part->_lo);
+    NativeType new_hi = MAX2(neg_part->_hi, pos_part->_hi);
+    assert(new_hi >= new_lo, "sanity");
+    return IntegerType::make(new_lo, new_hi, widen);
+  }
+
+  // Case B: divisor range does NOT span zero.
+  // Here i2 is entirely negative or entirely positive.
+  // Then i1/i2 is monotonic in i1 and i2 (when i2 keeps the same sign).
+  // Therefore the extrema occur at the four “corners”:
+  //   (i1->_lo, i2->_hi), (i1->_lo, i2->_lo), (i1->_hi, i2->_lo), (i1->_hi, i2->_hi).
+  // We compute all four and take the min and max.
+  // A special case handles overflow when dividing the most‐negative value by −1.
+
+  // adjust i2 bounds to not include zero, as zero always throws
+  NativeType i2_lo = i2->_lo == 0 ? 1 : i2->_lo;
+  NativeType i2_hi = i2->_hi == 0 ? -1 : i2->_hi;
+  constexpr NativeType min_val = std::numeric_limits<NativeType>::min();
+  static_assert(min_val == min_jint || min_val == min_jlong, "min has to be either min_jint or min_jlong");
+  constexpr NativeType max_val = std::numeric_limits<NativeType>::max();
+  static_assert(max_val == max_jint || max_val == max_jlong, "max has to be either max_jint or max_jlong");
+
+  // Special overflow case: min_val / (-1) == min_val (cf. JVMS§6.5 idiv/ldiv)
+  // We need to be careful that we never run min_val / (-1) in C++ code, as this overflow is UB there
+  if (i1->_lo == min_val && i2_hi == -1) {
+    NativeType new_lo = min_val;
+    NativeType new_hi;
+    // compute new_hi depending on whether divisor or dividend is non-constant.
+    // i2 is purely in the negative domain here (as i2_hi is -1)
+    // which means the maximum value this division can yield is either
+    // a) max_val           for non-constant dividend or
+    // b) (max_val / 2) + 1 for constant dividend and non-constant divisor or
+    // c) min_val           for constant dividend and constant divisor
+    if (!i1->is_con()) {
+      new_hi = max_val;
+      assert((min_val + 1) / -1 == new_hi, "new_hi should be max_val");
+    } else if (i2_lo != i2_hi) {
+      new_hi = (max_val / 2) + 1;
+      assert(min_val / -2 == new_hi, "new_hi should be (max_val / 2) + 1)");
+    } else {
+      new_hi = min_val;
+    }
```

Comment:
```suggestion
    if (!i1->is_con()) {
      // a) non-constant dividend: i1 could be min_val + 1.
      // -> i1 / i2 = (min_val + 1) / -1 = max_val is possible.
      new_hi = max_val;
      assert((min_val + 1) / -1 == new_hi, "new_hi should be max_val");
    } else if (i2_lo != i2_hi) {
      // b) i1 is constant min_val, i2 is non-constant.
      //    if i2 = -1 -> i1 / i2 =  min_val / -1 = min_val
      //    if i2 < -1 -> i1 / i2 <= min_val / -2 = (max_val / 2) + 1
      new_hi = (max_val / 2) + 1;
      assert(min_val / -2 == new_hi, "new_hi should be (max_val / 2) + 1)");
    } else {
      // c) i1 is constant min_val, i2 is constant -1.
      //    -> i1 / i2 = min_val / -1 = min_val
      new_hi = min_val;
    }
```
Reading this, I felt like I had to reconstruct a lot in my head. We could help the reader a little to close the gap.
Feel free to reformulate, and even to remove you list above in favour of the inlined comments.